core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -0,0 +1,259 @@
+            .ackTimeout(request.ackTimeout())
+            .masterNodeTimeout(request.masterNodeTimeout());
+        AliasAction[] actions = new AliasAction[2];
+        actions[0] = new AliasAction(AliasAction.Type.ADD, targetIndex, request.getSourceAlias());
+        actions[1] = new AliasAction(AliasAction.Type.REMOVE, concreteSourceIndex, request.getSourceAlias());
+        updateRequest.actions(actions);
+        return updateRequest;
+    }
+
+    static String generateRolloverIndexName(String sourceIndexName) {
+        int numberIndex = sourceIndexName.lastIndexOf("-");
+        int counter = 1;
+        String indexPrefix = sourceIndexName;
+        if (numberIndex != -1) {
+            try {
+                counter = Integer.parseInt(sourceIndexName.substring(numberIndex + 1));
+                counter++;
+                indexPrefix = sourceIndexName.substring(0, numberIndex);
+            } catch (NumberFormatException ignored) {
+            }
+        }
+        return String.join("-", indexPrefix, String.valueOf(counter));
+    }
+
+    static Set<Map.Entry<Condition, Boolean>> evaluateConditions(Set<Condition> conditions, long docCount, long indexAge) {
+        Set<Map.Entry<Condition, Boolean>> result = new HashSet<>(conditions.size());
+        for (Condition condition: conditions) {
+            if (condition instanceof Condition.MaxAge) {
+                Condition.MaxAge maxAge = (Condition.MaxAge) condition;
+                final TimeValue age = TimeValue.timeValueMillis(indexAge);
+                result.add(new AbstractMap.SimpleEntry<>(condition, maxAge.matches(age)));
+            } else if (condition instanceof Condition.MaxDocs) {
+                final Condition.MaxDocs maxDocs = (Condition.MaxDocs) condition;
+                result.add(new AbstractMap.SimpleEntry<>(condition, maxDocs.matches(docCount)));
+            } else {
+                throw new IllegalArgumentException("unknown condition [" + condition.getClass().getSimpleName() + "]");
+            }
+        }
+        return result;
+    }
+
+    static void validate(MetaData metaData, RolloverRequest request) {
+        final AliasOrIndex aliasOrIndex = metaData.getAliasAndIndexLookup().get(request.getSourceAlias());
+        if (aliasOrIndex == null) {
+            throw new IllegalArgumentException("source alias does not exist");
+        }
+        if (aliasOrIndex.isAlias() == false) {
+            throw new IllegalArgumentException("source alias is a concrete index");
+        }
+        if (aliasOrIndex.getIndices().size() != 1) {
+            throw new IllegalArgumentException("source alias maps to multiple indices");
+        }
+    }
+
+    static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(final String targetIndexName,
+                                                                          final RolloverRequest rolloverRequest) {
```

Comment:
Should we just serialize the entire source to construct the create index request later, as we don't know the name of the target index until we resolve the name of the underlying index?